lib/transforms/multipart.ts

```diff
@@ -25,70 +32,213 @@ function parseContentType(contentType?: string) {
+                    hk.notify();
+                    return;
+                }
+                // would be nice if Buffer had an indexOf. Would avoid a conversion to string.
+                // I could use node-buffertools but it introduces a dependency on a binary module.
+                const s = bbuf.toString('binary');
+                const ii = s.indexOf(boundary);
+                if (ii === 0) {
+                    const j = s.indexOf('\n', boundary.length);
+                    if (j < 0) throw new Error('newline missing after boundary');
+                    binReader.unread(bbuf.length - j - 1);
+                    hk.notify();
+                    return undefined;
+                } else if (ii > 0) {
+                    let j = s.lastIndexOf('\n', ii);
+                    if (s[j - 1] === '\r') j--;
+                    binReader.unread(bbuf.length - ii);
+                    return bbuf.slice(0, j);
+                } else {
+                    binReader.unread(bbuf.length - 31 * len);
+                    return bbuf.slice(0, 31 * len);
+                }
+            };
+            const partReader = generic.reader(read);
+            partReader.headers = headers;
+            writer.write(partReader);
+            hk.wait();
+        }
+    };
+}
+
+function mixedFormatter(ct: IMultipartContentType) {
+    const boundary = ct.boundary;
+    return (reader: Reader<Reader<Buffer>>, writer: Writer<Buffer>) => {
+        let part: Reader<Buffer> | undefined;
+        while ((part = reader.read()) !== undefined) {
+            const headers = part.headers;
+            if (!headers) throw new Error('part does not have headers');
+            Object.keys(part.headers).forEach(key => {
+                writer.write(Buffer.from(key + ': ' + headers[key] + '\n', 'binary'));
+            });
+            writer.write(Buffer.from('\n' + boundary + '\n'));
+            // cannot use pipe because pipe writes undefined at end.
+            part.forEach(data => {
+                writer.write(data);
+            });
+            writer.write(Buffer.from('\n' + boundary + '\n'));
+        }
+    };
+}
+
+function formDataParser(ct: IMultipartContentType): (reader: Reader<Buffer>, writer: Writer<any>) => void {
+    const boundary = '--' + ct.boundary;
+    return (reader: Reader<Buffer>, writer: Writer<any>) => {
+        const binReader = binary.reader(reader);
+        const hk = handshake();
+        while (true) {
+            const buf = binReader.readData(2048);
+            if (!buf || !buf.length) return;
+            const str = buf.toString('binary');
+
+            let endBoundaryIndex = str.indexOf(boundary) + boundary.length;
+            if (str.charAt(endBoundaryIndex) === '\r') {
+                endBoundaryIndex++;
+            }
+            if (str.charAt(endBoundaryIndex) === '\n') {
+                endBoundaryIndex++;
+            }
+            if (endBoundaryIndex < 0) throw new Error('boundary not found');
```

Comment:
shouldn't this test be just after `let endBoundaryIndex = ...`?
Actually, before the `+ boundary.length` increment.

Reply:
done